Rules/UseShouldProcessCorrectly.cs

```diff
@@ -299,48 +299,100 @@ private bool SupportsShouldProcess(string cmdName)
-                        if (cmdletBindingAttr != null)
+                        if (!functionInfo.CmdletBinding
+                            || functionInfo.ScriptBlock?.Attributes == null)
+                        {
+                            break;
+                        }
+
+                        foreach (CmdletBindingAttribute cmdletBindingAttribute in functionInfo.ScriptBlock.Attributes.OfType<CmdletBindingAttribute>())
                         {
-                            return cmdletBindingAttr.SupportsShouldProcess;
+                            return cmdletBindingAttribute.SupportsShouldProcess;
+                        }
+                    }
+                    catch
+                    {
+                        // functionInfo.ScriptBlock.Attributes may throw if it cannot resolve an attribute type
+                        // Instead we fall back to AST analysis
+                        // See: https://github.com/PowerShell/PSScriptAnalyzer/issues/1217
+                        if (TryGetShouldProcessValueFromAst(functionInfo, out bool hasShouldProcessSet))
+                        {
+                            return hasShouldProcessSet;
                         }
                     }
-                }
 
-                return false;
+                    break;
             }
 
-            var attributes = cmdletInfo.ImplementingType.GetTypeInfo().GetCustomAttributes(
-                typeof(System.Management.Automation.CmdletCommonMetadataAttribute),
-                true);
+            return false;
+        }
+
+        /// <summary>
+        /// Attempt to find whether a function has SupportsShouldProcess set based on its AST.
+        /// </summary>
+        /// <param name="functionInfo">The function info object referring to the function.</param>
+        /// <param name="hasShouldProcessSet">True if SupportsShouldProcess is set, false if not. Value is not valid if this method returns false.</param>
+        /// <returns>True if a value for SupportsShouldProcess was found, false otherwise.</returns>
+        private bool TryGetShouldProcessValueFromAst(FunctionInfo functionInfo, out bool hasShouldProcessSet)
+        {
+            // Get the body of the function
+            ScriptBlockAst functionBodyAst = (ScriptBlockAst)functionInfo.ScriptBlock.Ast.Find(ast => ast is ScriptBlockAst, searchNestedScriptBlocks: false);
 
-            foreach (var attr in attributes)
+            // Go through attributes on the parameter block, since this is where [CmdletBinding()] will be
+            foreach (AttributeAst attributeAst in functionBodyAst.ParamBlock.Attributes)
             {
-                var cmdletAttribute = attr as System.Management.Automation.CmdletAttribute;
-                if (cmdletAttribute != null)
+                // We're looking for [CmdletBinding()]
+                if (!String.Equals(attributeAst.TypeName.FullName, "CmdletBinding", StringComparison.OrdinalIgnoreCase))

```

Comment:
This is the same but simpler since `.Equals` is an extension method on `System.String`:
```suggestion
                if (!attributeAst.TypeName.FullName.Equals("CmdletBinding", StringComparison.OrdinalIgnoreCase))
```

Reply:
Actually the non-static `.Equals()` method [is just an instance method](https://github.com/dotnet/runtime/blob/a80e6d4c69fcfa1e17e17130c144a567b23a12a9/src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs#L604-L621) rather than an extension method. I generally try to use the [static `.Equals()` method](https://github.com/dotnet/runtime/blob/a80e6d4c69fcfa1e17e17130c144a567b23a12a9/src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs#L663-L677) because it has symmetric null checking (also the arguments are forced to be strings at compile time), although usually opt for whatever the standard in the codebase is.